### PR TITLE
Align KTO with DPO: Inline tokenization, new output format, DataCollatorForKTO

### PR DESCRIPTION
**Pull request by albertvillanova**
Align KTO with DPO: Inline tokenization, new output format, DataCollatorForKTO.

This PR consolidates the two-pass tokenization pipeline into a single pass and introduces a KTO-specific data collator.
- Subsequent PR: Move completion assembly from `_prepare_dataset` to `DataCollatorForKTO`

Part of:
- #4786

This PR refactors the KTO data preprocessing pipeline to simplify and clarify how tokenization, truncation, and padding are handled. The main changes move logic for assembling input and label tensors from the `_process_tokens` function into the main dataset mapping flow, introduce a new `DataCollatorForKTO` for padding, and streamline how KL datasets are assembled. The result is a more modular and maintainable codebase, with clearer separation of concerns between tokenization, truncation, and collation.

### Changes

**Delete `_process_tokens`:**
- The module-level `_process_tokens` function is removed. Its BOS/EOS reservation, truncation, and label-masking logic is now inlined directly into the `tokenize_fn` closure inside `_prepare_dataset`. This makes the preprocessing pipeline a single `.map` pass and easier to follow in one place.

**New output format from `tokenize_fn`:**
- `tokenize_fn` now produces `completion_input_ids`, `completion_attention_mask`, and `completion_labels` directly, instead of raw `answer_*` columns that required a second pass. The raw `answer_input_ids`/`answer_attention_mask` are still returned by `tokenize_fn` (untruncated) for the KL rotation step, and are dropped from the final dataset afterwards.

**Adapt KL assembly:**
- The KL processing path (`_process_tokens(prefix="KL_")`) is replaced by a new `assemble_kl_fn` closure. Since `prompt_input_ids` already carry BOS from `tokenize_fn`, `assemble_kl_fn` only needs to reserve budget for EOS. The intermediate `answer_input_ids`/`answer_attention_mask` columns are removed from the dataset after concatenation.

**Add `DataCollatorForKTO`:**
- Replaces `DPODataCollatorWithPadding` (imported from `experimental.utils`) with a KTO-specific `DataCollatorForKTO`. The new collator uses suffix-based dispatch (`_input_ids`, `_attention_mask`, `_labels`, `_logps`) and pads prompts on the left and completions on the right, handling both `completion_*` and `KL_completion_*` columns automatically.

**Test update:**
- Removes the `_process_tokens` import and its direct unit test. Updates the `_get_kl_dataset` rotation test to operate on a synthetic dataset, since `answer_input_ids` is no longer present in the final `trainer.train_dataset`.

**Documentation and logging:**
- Documentation strings and log messages have been updated to reference the new `DataCollatorForKTO` and the new data flow. 

These changes collectively make the KTO training pipeline more robust and easier to maintain, with clearer and more modular data handling.

---

> [!NOTE]
> **Medium Risk**
> Changes KTO dataset preprocessing, truncation, and padding paths (including KL dataset assembly), which can subtly affect training inputs/labels and loss behavior. Test updates reduce coverage of the removed `_process_tokens` path, so regressions may surface in edge-case tokenization/padding scenarios.
> 
> **Overview**
> KTO preprocessing is refactored to **assemble `completion_input_ids`, `completion_attention_mask`, and `completion_labels` inline during `_prepare_dataset` tokenization**, replacing the previous two-pass flow that relied on `_process_tokens`.
> 
> KL data preparation is updated to **rotate untruncated `answer_*` tokens with `_get_kl_dataset` and then assemble `KL_completion_*` via a new `assemble_kl_fn`**, followed by dropping intermediate `answer_*` columns.
> 
> A new `DataCollatorForUnpairedPreference` replaces `DPODataCollatorWithPadding` for KTO defaults, using suffix-based key handling and left-padding prompts/right-padding completions. Tests are updated to remove direct `_process_tokens` coverage and validate KL rotation using a synthetic dataset.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 5a1c4b22ab06d46bfa212f06934d91c07ad61f60. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/kto/kto_trainer.py [hunk 1]
```diff
@@ -91,82 +93,39 @@ def _get_kl_dataset(batch: dict[str, list[Any]]) -> dict[str, list[Any]]:
...
-def _process_tokens(example: dict[str, Any], model: "PreTrainedModel" = None, **kwargs) -> dict:
-    """Process tokens of a KTO specific dataset.
-
-    At this stage, we don't convert to PyTorch tensors yet; we just handle the truncation in case the prompt +
-    completion responses is/are too long. We truncate from the end (completion) to fit within max_length.
-
-    We also create the labels for the completion responses, which are of length equal to the sum of the length of the
-    prompt and the completion response, with `-100` for the prompt tokens.
+@dataclass
+class DataCollatorForKTO:
```
Comment (qgallouedec) [inline trl/experimental/kto/kto_trainer.py:97]: I’d go with `DataCollatorForUnpairedPreference`. Even if it’s closely associated with a specific method, the collator itself only operates on the data, so it’s technically method-agnostic. It also helps avoid slightly awkward patterns like the one we're removing in this PR, where a "DPO"-branded collator ends up being used for KTO. I'd aim for this convention `DataCollatorFor<DataType>`

**Comment by qgallouedec** [comment]
@codex review

File: trl/experimental/kto/kto_trainer.py [hunk 2]
```diff
@@ -91,82 +93,39 @@ def _get_kl_dataset(batch: dict[str, list[Any]]) -> dict[str, list[Any]]:
...
-    )
+    Args:
+        pad_token_id (`int`, defaults to `0`):
+            Token ID to use for padding `input_ids` sequences.
+    """
 
-    return batch
+    pad_token_id: int = 0
+
+    def __call__(self, features: list[dict[str, Any]]) -> dict[str, Any]:
```
Comment (qgallouedec) [inline trl/experimental/kto/kto_trainer.py:108]:
```suggestion
    def torch_call(self, features: list[dict[str, Any]]) -> dict[str, Any]:
```

**Review by qgallouedec — APPROVED** [review]
some suggestions, otherwise lgtm!
